### PR TITLE
http1: allow configuring format of encoded http1 response headers

### LINKED ISSUE #8463: In HTTP/1.1, envoy should pass-through header capitalization unmodified (without lower-casing)
*Title*: In HTTP/1.1, envoy should pass-through header capitalization unmodified (without lower-casing)

*Description*:
This ticket is related to https://github.com/envoyproxy/envoy/issues/6485, which was declined; with a renewed focus on the core point of discussion. HTTP/2 requires all header fields to be lower-cased. However, the [HTTP/1.1 standard](https://tools.ietf.org/html/rfc2616#section-4.2) is far less opinionated on the issue. On the one hand, it specifies that "field names are case-insensitive"; and on the other, it states:
```
Applications ought to follow "common form", where one is known or indicated,
when generating HTTP constructs, since there might exist some implementations
that fail to accept anything beyond the common forms.
```
Which could be interpreted to mean Train-Case as a default, since that is the form most commonly referred to in documentation of various headers (e.g. [Content-Type](https://www.w3.org/Protocols/rfc1341/4_Content-Type.html)).

Even if we take the standard to be ambiguous in this respect, there are major tools that deliver headers exactly as they were sent without modifying their case (e.g. [php request headers](https://www.php.net/manual/en/function.apache-request-headers.php)). This leads to real-world client code often being "broken" (in terms of API standard) by expecting headers in a specific casing.

While it may be reasonable to expect an organization to be able to internally fix its own client code, if Envoy is being used for APIs served externally to that organization, it will lead to avoidable -- sometimes show-stopping bugs. In our experience, we have found that trying to migrate to Envoy introduced a breaking change to our external APIs by hiding an optional header from certain client integrations (e.g. PHP). This failure case is silent and produces no errors on either side of the integration (uh-oh). As far as our client code is concerned, it's simply as if the header was never sent!

### PR DESCRIPTION
**Pull request by snowp**
Adds a configuration option that will convert all header keys into
Train-Case. This is useful to allow Envoy to respond with headers
that match the casing of other systems, to ensure that the wire format
of responses is unchanged when migrating to Envoy.

Signed-off-by: Snow Pettersen <snowp@squareup.com>

Risk Level: Mediun
Testing: UTs
Docs Changes: Proto comments
Release Notes: n/a
Fixes #8463

### CODE REVIEW

**Comment by snowp** [comment]
Mainly looking for API review at this point: is this too specific? Should we be aiming for something more flexible like arbitrary regex transformations with the ability to uppercase matches? 

For example, this could be represented by a series of regex matches: `{ match: /^[a-z]/, uppercase: true }` + `{ match: /[^0-9a-zA-Z][a-z]/, uppercase: true }`. This would come at some perf cost and code complexity but would make this much more flexible. 

Open to ideas for how to best express this so that it can be useful for as many people as possible.

@htuch @alyssawilk

File: api/envoy/api/v2/core/protocol.proto [hunk 1]
```diff
@@ -42,6 +42,24 @@ message Http1ProtocolOptions {
   // Envoy does not otherwise support HTTP/1.0 without a Host header.
   // This is a no-op if *accept_http_10* is not true.
   string default_host_for_http_10 = 3;
+
+  message HeaderKeyFormat {
+    enum HeaderFormat {
```
Comment (mattklein123) [inline api/envoy/api/v2/core/protocol.proto]: enums are pretty fragile for stuff like this. I would recommend just making this a oneof with a required value and a single `TrainCase` message that is empty. In the future we can easily add configuration to the TrainCase message or add more complicated message types.

**Comment by htuch** [comment]
@snowp would pass-thru of client casing work to support this use case? @jmarantz has done some research on this and it seems doable.

**Comment by snowp** [comment]
Pass-through would work in certain scenarios (h/1.1 all the way), but I had some concerns about supporting intermediate h2 hops, as we tend to use h2 wherever possible. Seems like @alyssawilk shared some of those concerns based on the conversation in #8463.

File: api/envoy/api/v2/core/protocol.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/api/v2/core/protocol.proto]: That's probably a good idea.

File: api/envoy/api/v2/core/protocol.proto [hunk 2]
```diff
@@ -42,6 +42,24 @@ message Http1ProtocolOptions {
...
+      // By default, all headers are lower cased.
+      DEFAULT = 0;
+
+      // Formats the header using Train-Case: the first character and any character following a
+      // special character will be capitalized if it's an alpha character. For example,
+      // "content-type" becomes "Content-Type", and "foo$b#$ar" becomes "Foo$B#$Ar".
+      TRAIN_CASE = 1;
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: Is "train case" a standard term? I haven't heard it before.

File: api/envoy/api/v2/core/protocol.proto [hunk 3]
```diff
@@ -42,6 +42,24 @@ message Http1ProtocolOptions {
...
+    }
+
+    HeaderFormat header_format = 1;
+  }
+
+  // Describes how the keys for response headers should be formatted. By default, all header keys
+  // are lower cased.
+  HeaderKeyFormat response_header_key_format = 4;
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: Is this only for response? What about request headers forwarded to the upstream?

**Comment by htuch** [comment]
@snowp ack, let's continue with explicit capitalization support then.

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: i hadn't heard about it either before it was used in the issue that was filed. Googling around I can find some references to it (https://docs.rs/Inflector/0.7.0/inflector/cases/traincase/index.html, https://medium.com/infancyit/letter-case-special-styles-7fa369bf173d), though other sources have it mean something else (https://en.wikipedia.org/wiki/Letter_case#Special_case_styles says TRAIN-CASE, https://www.fluentcpp.com/2019/01/08/restmycase-a-c-library-for-formatting-string-cases/ says train_Case).

Open to naming suggestions here, given that this doesn't seem  to be a very common term

Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: One option might be `rfc7230_casing` since we're basically trying to present the headers the way they are cased in the HTTP/1.1 RFCs.

File: api/envoy/api/v2/core/protocol.proto (hunk 3 above)
Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: i can add it for the request path too, but that seems to be a bit more involved since the http1 codec settings aren't currently being passed through to the connection pool/upstream http/1.1 codec (since all the current options apply only to the HCM). 

How about i add the proto def /w not-implemented and leave that as a TODO?

Comment (alyssawilk) [reply api/envoy/api/v2/core/protocol.proto]: I think it makes sense to have it just be header_format, and keep it hidden until we have it implemented on both sides, because being allowd to set response header format where the settings are used for requests is weird.
WDYT?

Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: yeah i like that, will go with that

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/api/v2/core/protocol.proto]: that sounds good to me. @PiotrSikora any concerns about taking the name of the RFC in vain?

File: api/envoy/api/v2/core/protocol.proto (hunk 3 above)
Comment (htuch) [reply api/envoy/api/v2/core/protocol.proto]: Arguably though clients and backends might have different case senstivity concerns. E.g. one wants Rfc7230-Casing, the other wants 1337Spe4k-k451nG.

Comment (alyssawilk) [reply api/envoy/api/v2/core/protocol.proto]: Sure, but we have codec options in both places.
AFIK when there's protocol options on an HCM, this will affect how we serialize the response headers and when we set the protocol options in CDS it'll affect how we serialize request headers. Since we're not doing enforcement of request headers at the listener or response headers in the connection pool, I don't know what it would mean to have response headers in the connection pool Http1 options config.

I suggest we just have one name, and note it affects how we serialize the headers out at the codec.

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (PiotrSikora) [reply api/envoy/api/v2/core/protocol.proto]: `"TrainCase"` is called [`PascalCase (aka Upper Camel Case)`](https://en.wikipedia.org/wiki/Pascal_case).

We should also have the `DONT_MODIFY` case (probably a good default, and in theory it's backwards compatible, since HTTP headers are case-insensitive), because not all standard headers use `PascalCase` (e.g. `Content-MD5` and `TE`), and use transcoding to `PascalCase` or `lowercase` only when transforming headers between HTTP/1.1 and HTTP/2.

Comment (PiotrSikora) [reply api/envoy/api/v2/core/protocol.proto]: Nevermind, PascalCase is without hyphens.

Comment (PiotrSikora) [reply api/envoy/api/v2/core/protocol.proto]: @htuch using RFC number where isn't appropriate here, IMHO, because that would imply that the casing in defined in that RFC, whereas that's not the case (as far as I can tell).

File: api/envoy/api/v2/core/protocol.proto (hunk 3 above)
Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: Yeah, I think the fact that the configuration for request and response headers wouldn't occur in the same location (HCM vs Cluster) makes having it be a single field make more sense than two fields with only one ever being relevant nicer

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: How about `propercase_words`? then we define a word as any sequence of alpha characters

**Comment by snowp** [comment]
This should be ready for review for whoever wants to pick this up

File: api/envoy/api/v2/core/protocol.proto (hunk 3 above)
Comment (htuch) [reply api/envoy/api/v2/core/protocol.proto]: Good point, a single name makes sense given the multiplicity of codec options on downstream/upstream.

File: api/envoy/api/v2/core/protocol.proto [hunk 4]
```diff
@@ -50,6 +62,11 @@ message Http1ProtocolOptions {
   // Envoy does not otherwise support HTTP/1.0 without a Host header.
   // This is a no-op if *accept_http_10* is not true.
   string default_host_for_http_10 = 3;
+
+  // Describes how the keys for response headers should be formatted. By default, all header keys
+  // are lower cased.
+  // [#not-implemented-hide:] Hiding until both request and response formating has been implemented.
```
Comment (alyssawilk) [inline api/envoy/api/v2/core/protocol.proto]: Any reason to not just do it here?  Is it more work than creating the formatter on the client codec?
(I'm totally fine with a follow-up if you prefer)

File: source/common/http/http1/header_formatter.cc [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#include "common/http/http1/header_formatter.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+std::string ProperCaseHeaderKeyFormatter::format(absl::string_view key) const {
```
Comment (alyssawilk) [inline source/common/http/http1/header_formatter.cc]: Looking at https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

This won't do A-IM, TE Content-MD5, HTTP2-Setting, and about 10 others the way an HTTP/1 user might expect.

I'd lean towards saying we do standard capitalization of standard HTTP 1.1 headers that Envoy knows about (e.g. TE) with a link to headers Envoy knows about.  ProperCaseExceptForRegisteredStandardizedHeaderExceptions is too wordy, but I think that's what we want?

This way if someone later on desperately needs P3P cased they awy they want, they can just add it to the list of Envoy known headers.  And when we eventually add the Envoy extension mechanism for arbitrary headers we keep talking about, folks can do their own custom builds with TheirWhackyCAPs if they want.

Alternately if we want to go for the simple case, I think we should document there are known standard headers which will still look unusual to HTTP/1 consumers, and someone can implement ProperCaseByDefaultButWithOverrides later now we have an API for things.

@mattklein123 WDYT?

File: tools/spelling_dictionary.txt [hunk 6]
```diff
@@ -310,6 +310,7 @@ WS
 Welford's
 Werror
 XDS
+ar
```
Comment (alyssawilk) [inline tools/spelling_dictionary.txt]: can we change foo$b#$ar to foo$b#$are for all the various explantions of that transform to avoid this addition?  Feels like it'd be a common typo of "are" which we should catch.

File: api/envoy/api/v2/core/protocol.proto (hunk 4 above)
Comment (snowp) [reply api/envoy/api/v2/core/protocol.proto]: Adding the codec stuff is easy, the thing i was trying to avoid was doing all the plumbing + tests to make the h1 protocol options available to the connection pool. I'll do a fast follow once this lands

File: source/common/http/http1/header_formatter.cc (hunk 5 above)
Comment (snowp) [reply source/common/http/http1/header_formatter.cc]: Maybe `StandardOrProperCase`? I can go either way on whether we want to handle well known headers, it sounds a tad bit more in line with what someone might expect but it might be overkill.

Looking at Go's net/http (https://golang.org/src/net/http/header.go#L214) they do the same kind of formatting this PR does when canonicalizing headers except that they only uppercase letters after a hyphen.

File: api/envoy/api/v2/core/protocol.proto [hunk 7]
```diff
@@ -34,6 +34,18 @@ message HttpProtocolOptions {
 }
 
 message Http1ProtocolOptions {
+  message HeaderKeyFormat {
+    message ProperCaseWords {
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply api/envoy/api/v2/core/protocol.proto]: Hah, I was about to ask w.r.t. the general HTTP/1->HTTP/3 I asked dan to add - didn't know if we wanted to move to a oneof as we moved it to a general place.  I think I hit our first "trying to deprecated = 0" issue so I can get behind this - just wanted to make sure I knew for future API reviews.

File: api/envoy/api/v2/core/protocol.proto [hunk 8]
```diff
@@ -34,6 +34,18 @@ message HttpProtocolOptions {
...
+    }
+
+    oneof header_format {
```
Comment (mattklein123) [inline api/envoy/api/v2/core/protocol.proto]: Please add a required validation for this oneof.

File: source/common/http/http1/codec_impl.cc [hunk 9]
```diff
@@ -67,7 +70,7 @@ void StreamEncoderImpl::encode100ContinueHeaders(const HeaderMap& headers) {
 void StreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream) {
   bool saw_content_length = false;
   headers.iterate(
-      [](const HeaderEntry& header, void* context) -> HeaderMap::Iterate {
+      [this](const HeaderEntry& header, void* context) -> HeaderMap::Iterate {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: FWIW, this function used a C-style callback originally for performance reasons when I was optimizing for the horse race case. I would leave a comment here about looking at perf, or potentially just pre-create the lambda at construction time so that in the keep alive case we don't need any extra allocations.

File: source/common/http/http1/header_formatter.cc (hunk 5 above)
Comment (mattklein123) [reply source/common/http/http1/header_formatter.cc]: I don't have a strong opinion here as long as it is well documented. The way that we have things set up at the API layer we can always make things configurable later if needed.

File: api/envoy/api/v2/core/protocol.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/api/v2/core/protocol.proto]: I think for naming protocols, an enum is fine. These are scalar values. The issue here is that there may be more structured data that we want to support internally within each value; this is like with LB in clusters, you have both an enum to pick the LB and then a bunch of random fields that apply to each of the different LB values. The oneof+messages pattern is much nicer for that situation.

File: api/envoy/api/v2/core/protocol.proto [hunk 10]
```diff
@@ -34,6 +34,18 @@ message HttpProtocolOptions {
...
+      // Formats the header by proper casing words: the first character and any character following
+      // a special character will be capitalized if it's an alpha character. For example,
+      // "content-type" becomes "Content-Type", and "foo$b#$ar" becomes "Foo$B#$Ar".
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: @envoyproxy/api-shepherds API LGTM. I think if there are nuances around details with how this is done, e.g. whether it's `-` only or any punctuation, we can sort them out with `ProperCaseWords` fields later.

File: source/common/http/http1/codec_impl.cc (hunk 9 above)
Comment (snowp) [reply source/common/http/http1/codec_impl.cc]: Do you think it's worth having multiple versions of iterate, one with C-style callback and one with std::function? afaik std::function adds some overhead compared to C-style callbacks due to the additional indirection

File: source/common/http/http1/header_formatter.cc (hunk 5 above)
Comment (rkofman) [reply source/common/http/http1/header_formatter.cc]: A soft suggestion of having `StandardOrProperCase` behavior; and perhaps condensing the config name down to `StandardCase` (which would follow well-known casings for well-known headers, and standard casing patterns for H/1 otherwise).

File: source/common/http/http1/codec_impl.cc (hunk 9 above)
Comment (snowp) [reply source/common/http/http1/codec_impl.cc]: Turns out I was able to do this without changing the signature, since I could just use the encoder that was already being passed as the context to resolve the formatter.

File: include/envoy/http/codec.h [hunk 11]
```diff
@@ -229,6 +229,19 @@ struct Http1Settings {
   bool accept_http_10_{false};
   // Set a default host if no Host: header is present for HTTP/1.0 requests.`
   std::string default_host_for_http_10_;
+
+  enum class HeaderKeyFormat {
+    // By default no formatting is performed, presenting all headers in lowercase (as Envoy
+    // internals normalize everything to lowercase.)
+    Default,
+    // Performs proper casing of header keys: the first and all alpha characters following a
+    // non-alphanumeric
```
Comment (mattklein123) [inline include/envoy/http/codec.h]: nit: reflow comment.

File: api/envoy/api/v3alpha/core/protocol.proto [hunk 12]
```diff
@@ -51,6 +65,11 @@ message Http1ProtocolOptions {
   // Envoy does not otherwise support HTTP/1.0 without a Host header.
   // This is a no-op if *accept_http_10* is not true.
   string default_host_for_http_10 = 3;
+
+  // Describes how the keys for response headers should be formatted. By default, all header keys
+  // are lower cased.
+  // [#not-implemented-hide:] Hiding until both request and response formating has been implemented.
```
Comment (mattklein123) [inline api/envoy/api/v3alpha/core/protocol.proto]: So you will do docs / release notes in the follow up change?

File: source/common/http/http1/codec_impl.h [hunk 13]
```diff
@@ -96,18 +97,22 @@ class StreamEncoderImpl : public StreamEncoder,
    */
   void endEncode();
 
+void encodeFormattedHeader(absl::string_view key, absl::string_view value);
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.h]: Can you get rid of the clang-format-off/on above, as macro escaping is turning the formatter off here. (and we don't do that any more for the stat macros)

File: source/common/http/http1/header_formatter.cc [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  auto copy = std::string(key);
+
+  bool shouldCapitalize = true;
```
Comment (mattklein123) [inline source/common/http/http1/header_formatter.cc]: nit: snake_case

File: test/common/http/http1/header_formatter_test.cc [hunk 15]
```diff
@@ -0,0 +1,18 @@
+#include "common/http/http1/header_formatter.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+TEST(ProperCaseHeaderKeyFormatterTest, Formatting) {
```
Comment (mattklein123) [inline test/common/http/http1/header_formatter_test.cc]: Can we significantly bump coverage here including empty values, single character values, etc.

File: api/envoy/api/v3alpha/core/protocol.proto (hunk 12 above)
Comment (snowp) [reply api/envoy/api/v3alpha/core/protocol.proto]: yep

File: api/envoy/api/v2/core/protocol.proto [hunk 16]
```diff
@@ -35,6 +35,20 @@ message HttpProtocolOptions {
...
+    message ProperCaseWords {
+    }
+
+    oneof header_format {
+      option (validate.required) = true;
+
+      // Formats the header by proper casing words: the first character and any character following
+      // a special character will be capitalized if it's an alpha character. For example,
+      // "content-type" becomes "Content-Type", and "foo$b#$are" becomes "Foo$B#$Are".
+      ProperCaseWords proper_case_words = 1;
```
Comment (alyssawilk) [inline api/envoy/api/v2/core/protocol.proto]: Optional: Do you think it's worth calling out both that this is an effort to normalize HTTP/1 headers to what users expect, and that this will not behave "as expected" for all headers, e.g. TE?  I can imagine that folks who have HTTP/1 casing issues will think this is the solution to their problem and I think documenting the simplicity vs the non-handled corner cases will probably save someone some trouble.

File: source/common/http/http1/codec_impl.cc [hunk 17]
```diff
@@ -116,17 +129,14 @@ void StreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream)
...
       }
       chunk_encoding_ = false;
     } else if (connection_.protocol() == Protocol::Http10) {
       chunk_encoding_ = false;
     } else {
-      encodeHeader(Headers::get().TransferEncoding.get().c_str(),
-                   Headers::get().TransferEncoding.get().size(),
-                   Headers::get().TransferEncodingValues.Chunked.c_str(),
-                   Headers::get().TransferEncodingValues.Chunked.size());
+      encodeFormattedHeader(Headers::get().TransferEncoding.get().c_str(),
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: can we skip the c_str() here and in other places?

Comment (snowp) [reply source/common/http/http1/codec_impl.cc]: done

**Comment by mattklein123** [comment]
Please merge master to fix coverage.

/wait

File: source/common/http/http1/codec_impl.cc [hunk 18]
```diff
@@ -34,10 +36,32 @@ const StringUtil::CaseUnorderedSet& caseUnorderdSetContainingUpgradeAndHttp2Sett
...
 
-StreamEncoderImpl::StreamEncoderImpl(ConnectionImpl& connection) : connection_(connection) {
+StreamEncoderImpl::StreamEncoderImpl(ConnectionImpl& connection,
+                                     HeaderKeyFormatter* header_key_formatter)
+    : connection_(connection), header_key_formatter_(header_key_formatter) {
   if (connection_.connection().aboveHighWatermark()) {
     runHighWatermarkCallbacks();
   }
+
+  encode_header_cb_ = [](const HeaderEntry& header, void* context) -> HeaderMap::Iterate {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: Sorry I think we can kill this member variable and revert this part of the change now, right?

File: source/common/http/http1/codec_impl.h [hunk 19]
```diff
@@ -96,18 +94,23 @@ class StreamEncoderImpl : public StreamEncoder,
...
   void endEncode();
 
+  void encodeFormattedHeader(absl::string_view key, absl::string_view value);
+
   bool chunk_encoding_{true};
   bool processing_100_continue_{false};
   bool is_response_to_head_request_{false};
   bool is_content_length_allowed_{true};
+  HeaderMap::ConstIterateCb encode_header_cb_;
+  HeaderKeyFormatter* header_key_formatter_;
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.h]: nit: can this be const? (Both the pointer and the object)

File: source/common/http/http1/header_formatter.cc (hunk 14 above)
Comment (mattklein123) [reply source/common/http/http1/header_formatter.cc]: bump

**Comment by sunnoy** [comment]
So how to configure the custom header format, for example, I want to get this format header：

```bash
Sec-WebSocket-Accept

not 

Sec-Websocket-Accept
```
